File: app/src/main/java/de/grobox/transportr/ui/LineView.java
/*    Transportr
 *    Copyright (C) 2013 - 2016 Torsten Grote
 *
 *    This program is Free Software: you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as
 *    published by the Free Software Foundation, either version 3 of the
 *    License, or (at your option) any later version.
 *
 *    This program is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package de.grobox.transportr.ui;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.GradientDrawable;
import android.support.v4.content.ContextCompat;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import de.grobox.transportr.R;
import de.schildbach.pte.dto.Line;

import static android.content.Context.LAYOUT_INFLATER_SERVICE;
import static de.grobox.transportr.utils.TransportrUtils.getDrawableForProduct;

public class LineView extends LinearLayout {

	private final ViewHolder ui;

	public LineView(Context context, AttributeSet attr) {
		super(context, attr);

		setOrientation(HORIZONTAL);

		//noinspection deprecation
		setBackgroundDrawable(ContextCompat.getDrawable(context, R.drawable.line_box));

		int size = context.getResources().getDimensionPixelSize(R.dimen.line_box_height);
		setMinimumHeight(size);
		setMinimumWidth(size);

		LayoutInflater inflater = (LayoutInflater) context.getSystemService(LAYOUT_INFLATER_SERVICE);
		inflater.inflate(R.layout.line_view, this, true);
		ui = new ViewHolder(this);
	}

	public LineView(Context context) {
		this(context, null);
	}

	public void setLine(Line line) {
		if (line.product != null) {
			Drawable drawable = ContextCompat.getDrawable(getContext(), getDrawableForProduct(line.product));
			ui.product.setImageDrawable(drawable);
		} else {
			ui.product.setVisibility(GONE);
		}

		ui.label.setText(line.label);
		ui.label.setVisibility(VISIBLE);

		if (line.style != null) {
			GradientDrawable box = (GradientDrawable) ui.box.getBackground();

			if(box != null) {
				// change color and mutate before to not share state with other instances
				box.mutate();
				box.setColor(line.style.backgroundColor);
			}
			ui.label.setTextColor(line.style.foregroundColor);
			ui.product.setColorFilter(line.style.foregroundColor);
		}
	}

	public void setWalk() {
		ui.product.setImageResource(R.drawable.ic_walk);
		ui.label.setVisibility(GONE);
		ui.box.setBackgroundResource(R.drawable.walk_box);
	}

	public String getLabel() {
		return (String) ui.label.getText();
	}

	private static class ViewHolder {
		private ViewGroup box;
		private ImageView product;
		private TextView label;

		public ViewHolder(View view) {
			box = (ViewGroup) view;
			product = view.findViewById(R.id.productView);
			label = view.findViewById(R.id.labelView);
		}
	}

}

File: app/src/main/java/de/grobox/transportr/ui/LineView.kt
/*
 *    Transportr
 *
 *    Copyright (c) 2013 - 2021 Torsten Grote
 *
 *    This program is Free Software: you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as
 *    published by the Free Software Foundation, either version 3 of the
 *    License, or (at your option) any later version.
 *
 *    This program is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package de.grobox.transportr.ui

import android.content.Context
import android.graphics.PorterDuff.Mode.SRC_IN
import android.graphics.drawable.GradientDrawable
import androidx.annotation.ColorInt
import androidx.annotation.DrawableRes
import androidx.core.content.ContextCompat.getColor
import androidx.core.content.ContextCompat.getDrawable
import androidx.appcompat.widget.AppCompatTextView
import android.util.AttributeSet
import de.grobox.transportr.R
import de.grobox.transportr.utils.TransportrUtils.dpToPx
import de.grobox.transportr.utils.TransportrUtils.getDrawableForProduct
import de.schildbach.pte.dto.Line

class LineView(context: Context, attr: AttributeSet?) : AppCompatTextView(context, attr) {

    fun setLine(line: Line) {
        // get colors
        val foregroundColor = line.style?.foregroundColor
        val backgroundColor = line.style?.backgroundColor

        // set colored background
        val backgroundDrawable = getDrawable(context, R.drawable.line_box) as GradientDrawable
        backgroundDrawable.mutate() // mutate to not share state with other instances
        if (backgroundColor != null) backgroundDrawable.setColor(backgroundColor)
        @Suppress("DEPRECATION")
        setBackgroundDrawable(backgroundDrawable)

        // correct padding that gets lost when setting background
        setPadding(0, dpToPx(context, 2), dpToPx(context, 4), dpToPx(context, 2))

        // product icon
        if (line.product != null) {
            setDrawable(getDrawableForProduct(line.product), foregroundColor)
        }

        // set colored label
        text = line.label
        if (foregroundColor != null) setTextColor(foregroundColor)
    }

    fun setWalk() {
        setDrawable(R.drawable.ic_walk, getColor(context, android.R.color.black))
        setBackgroundResource(R.drawable.walk_box)
    }

    private fun setDrawable(@DrawableRes res: Int, @ColorInt color: Int?) {
        val drawable = getDrawable(context, res)!!
        color?.let {
            drawable.mutate().setColorFilter(color, SRC_IN)
        }
        setCompoundDrawablesWithIntrinsicBounds(drawable, null, null, null)
    }

}
